statie/source/_posts/2019/2019-05-21-review-of-the-refactoring.md

```diff
@@ -37,10 +37,62 @@ Almost all refactorings include inverse refactoring.
 When there is [Extract Variable](https://refactoring.com/catalog/extractVariable.html), you can find also [Inline Variable](https://refactoring.com/catalog/inlineVariable.html).
 That was a surprise for me as I thought the first option is always better (and I learned that's not true).
 
+Extracting variable gives a name to the statement
+```php
+return preg_replace('/(?:[,.]0+|([,.]\d*?)0+)$/', '$1', $value);
+
+// refactored to
+$croppingZerosPattern = '/(?:[,.]0+|([,.]\d*?)0+)$/';
+return preg_replace($croppingZerosPattern, '$1', $value);
+```
+
+On the other hand when the statement is self-explanatory, we can inline the variable
+```php
+$itemsPrice = $this->getItemsPrice();
+if ($itemsPrice > $this->discountTreshold) {
+
+// refactored to
+if ($this->getItemsPrice() > $this->discountTreshold) {
+```
+
 I want to point out [Replace Type Code with Subclasses](https://refactoring.com/catalog/replaceTypeCodeWithSubclasses.html) and [Introduce Special Case](https://refactoring.com/catalog/introduceSpecialCase.html).
 Both introduce reasonable inheritance.
 Reasonable inheritance is the most difficult part of OOP for me and these refactorings taught me something new.
 
+Calculating coupon discount refactored from anemic model to objects using inheritance
+```php
+if ($coupon->getType() === Coupon::PERCENTAGE) {
+    $cart = $coupon->applyPercentaceDiscount($cart);
+} elseif ($coupon->getType() === Coupon::ABSOLUTE) {
+    $cart = $coupon->applyAbsoluteDiscount($cart);
+}
+
+// refactored to
+abstract class Coupon {
+    abstract public function applyDiscount(Cart $cart): Cart;
+}
+final class PercentageCoupon extends Coupon {...}
+final class AbsoluteCoupon extends Coupon {...}
+
+$cart = $coupon->applyDiscount($cart);
```

Comment:
Collector pattern :heart: 

Personally I know what's inside `AbsoluteCoupon`, but for the reader it must be magic - "you create 2 classes, but what happens with the logic"?

Could you share its content?

Reply:
is it better? I don't want to attract reader's focus to a logic, I want to show only the concept - inheritance

Reply:
Much better, now I get it :+1: